W-19316812-neo4j-major-documentation-kt (#2974)
mulesoft/docs-connectors: neo4j/4.0/modules/ROOT/pages/neo4j-connector-upgrade-migrate.adoc

= Upgrading and Migrating Neo4j Connector to Version 4.0.0

Upgrade Anypoint Connector for Neo4j (Neo4j Connector) to version 4.0.0.

== Supported Upgrade Paths

[%header,cols="50a,50a"]
|===
|From Version | To Version
|3.0.x |4.0.0
|===

== Changes in This Release

This release contains significant changes:

> * Upgraded the Neo4j Java driver from version 1.7.6 to 5.28.8.

Reviewer: (Globally) bullet points should not end with a period or full stop

> * Updated Neo4j server requirement to version 4.4+ (previously supported Neo4j 3.x).

Reviewer: ```suggestion
* Updated Neo4j server requirements to version 4.4 and later (previously supported Neo4j 3.x).
```

* Required minimum JDK 17 (upgraded from JDK 8).
* Updated minimum Mule runtime engine (Mule) to 4.9.0 (upgraded from 4.1.x).
* Removed the *REST URL* parameter from the connector configuration.
* Overhauled input and output metadata resolution for a better Anypoint Studio experience.
* Provided full compatibility with Neo4j 5.x and 2025.x server features.
* Gained access to reactive APIs, multi-database support, and enhanced security.
* Updated Docker configuration settings to use current Neo4j standards.
* Changed the connector to only require a BOLT URL configuration.
* Changed all metadata operations to use native Cypher queries through a BOLT connection instead of REST API calls.

== Requirements

Ensure that you meet these requirements before you perform the upgrade.

=== Neo4j Server Requirements

> * The minimum Neo4j server version required is 4.4 LTS or higher.

Reviewer: ```suggestion
* The minimum Neo4j server version required is 4.4 LTS or later.
```

* The recommended versions are Neo4j 5.26 LTS or Neo4j 2025.06+.
* Neo4j 3.x servers are no longer supported.

=== Java Runtime Requirements

> * The minimum JDK version required is 17

Reviewer: (Globally) be consistent with the period or full stop.

= Upgrading and Migrating Neo4j Connector to Version 4.0.0

Upgrade Anypoint Connector for Neo4j (Neo4j Connector) to version 4.0.0.

== Supported Upgrade Paths

[%header,cols="50a,50a"]
|===
|From Version | To Version
|3.0.x |4.0.0
|===

== Changes in This Release

This release contains significant changes:

* Upgraded the Neo4j Java driver from version 1.7.6 to 5.28.8.
* Updated Neo4j server requirement to version 4.4+ (previously supported Neo4j 3.x).
* Required minimum JDK 17 (upgraded from JDK 8).
* Updated minimum Mule runtime engine (Mule) to 4.9.0 (upgraded from 4.1.x).
* Removed the *REST URL* parameter from the connector configuration.
* Overhauled input and output metadata resolution for a better Anypoint Studio experience.
* Provided full compatibility with Neo4j 5.x and 2025.x server features.
* Gained access to reactive APIs, multi-database support, and enhanced security.
* Updated Docker configuration settings to use current Neo4j standards.

> * The connector now only requires a BOLT URL configuration.

Reviewer 2: ```suggestion
* Changed the connector to only require a BOLT URL configuration.
```

> * All metadata operations use native Cypher queries through a BOLT connection instead of REST API calls.

Reviewer 2: ```suggestion
* Changed all metadata operations to use native Cypher queries through a BOLT connection instead of REST API calls.
```

== Requirements

> Ensure that you meet these requirements before you perform the upgrade:

Reviewer 2: ```suggestion
Ensure that you meet these requirements before you perform the upgrade.
```

Reviewer 2: Don't introduce sections with a colon.

=== Neo4j Server Requirements

> * The minimum Neo4j server version required is 4.4 LTS or higher

Reviewer 2: Because these are all full sentences, add periods.

* The recommended versions are Neo4j 5.26 LTS or Neo4j 2025.06+
* Neo4j 3.x servers are no longer supported

=== Java Runtime Requirements

* The minimum JDK version required is 17.
* The only supported JDK version is 17.
* JDK 8 and 11 are no longer supported.

=== Mule Requirements

> 

Reviewer 2: These aren't complete sentences, so correct as is.

* Minimum Mule version 4.9.0, required for JDK 17 support and unified connectors parent compatibility
* Mule versions 4.1.x - 4.8.x not supported

=== Development Environment

* Anypoint Studio latest version supporting Mule Runtime 4.9.0+
* Maven version supporting JDK 17+

> == Limitations

Reviewer 2: Maybe this should be Considerations?

Before you perform the upgrade, be aware of these known limitations:

> * Manually update all Mule applications to support new requirements

Reviewer 2: This is a prereq, not a limitation.

Look at this list and make parallel. Also, add periods since they're full sentences.

* After the connector is upgraded, rolling back requires restoring from backup
* Neo4j 3.x servers cannot be used with connector version 4.0.0
* Mule applications must run on Mule 4.9.0+ and JDK 17+
* Update existing configurations using REST URL parameter

> * After upgrading the connector, re-test all existing connections

Reviewer 2: Also not a limitation.

* Refresh Studio metadata for the connector operations
* Maven builds require JDK 17+ toolchain
* Upgrade test servers to Neo4j version 4.4 + before performing connector testing

== Upgrade Prerequisites

> Before you perform the upgrade, you must complete the following:

Reviewer 2: colon

> === Environment Preparation

Reviewer 2: Can you make these imperative, since they're instructions to the user?

Prepare Your Environment

> * Create a backup of your Mule applications, data, and configurations

Reviewer 2: Should these be an ordered list?

Add periods.

> * Upgrade Neo4j Server to version 4.4 LTS or higher (if currently using Neo4j 3.x)

Reviewer 2: Use "or later" per the style guidelines.

* Upgrade Mule to version 4.9.0 or higher
* Upgrade JDK to version 17 or higher on all environments (development, test, production)

> === Mule Application Assessment

Reviewer 2: Assess Mule Applications

> * Review existing configurations to identify REST URL usage

Reviewer 2: Numbered list and periods

* Document current BOLT URL settings for post-upgrade validation
* Test connectivity to make sure Neo4j servers are accessible via BOLT protocol

> * Identify custom error handling that may reference connector-specific error types

Reviewer 2: ```suggestion
* Identify custom error handling that references connector-specific error types.
```

. In *Project Properties*, select *Connectors*.
. Select the latest version for the connector, and click *Apply*.

== Upgrade the Connector in Studio

Follow these steps to perform the upgrade:

. In Anypoint Studio, open your existing Mule project.
. In *Mule Palette*, click *Search in Exchange*.
. In *Add Dependencies to Project*, enter `neo4j` in the search field.
. In *Available modules*, select *Neo4j Connector* and click *Add*.
. Click *Finish*.
. Verify that the Neo4j Connector `<dependency>` version is `4.0.0` in the `pom.xml` file in the Mule project:
+
[source,xml,linenums]
----
<dependency>
  <groupId>com.mulesoft.connectors</groupId>
  <artifactId>mule-neo4j-connector</artifactId>
  <version>4.0.0</version>
  <classifier>mule-plugin</classifier>
</dependency>
----

Studio upgrades the connector automatically, but configuration changes are required.

== Post Upgrade Steps

After you install the latest version of the connector, follow these steps to complete the upgrade:

=== Remove REST URL Parameter

The REST URL parameter is no longer supported in the connector configuration. Update your configurations:

> Before (Version 3.x):

Reviewer 2: You could use a DL here to get bold and an easier to scan format.

[source,xml,linenums]
----
<neo4j:config name="Neo4j_Config" doc:name="Neo4j Config">
    <neo4j:connection username="${config.username}"
                     password="${config.password}"
                     boltUrl="${config.boltUrl}"
                     restUrl="${config.restUrl}" />
</neo4j:config>
----

After (Version 4.0.0):
[source,xml,linenums]
----
<neo4j:config name="Neo4j_Config" doc:name="Neo4j Config">
    <neo4j:connection username="${config.username}"
                     password="${config.password}"
                     boltUrl="${config.boltUrl}" />
</neo4j:config>
----

=== Update Connection Properties

Remove the REST URL property from your configuration files:

> Before:

Reviewer 2: Same as previous comment

[source,text,linenums]
----
config.username=<USERNAME>
config.password=<PASSWORD>
config.boltUrl=<BOLT_URL_ENDPOINT>
config.restUrl=<REST_URL_ENDPOINT>  <!-- Remove this line -->
----

After:
[source,text,linenums]
----
config.username=<USERNAME>
config.password=<PASSWORD>
config.boltUrl=<BOLT_URL_ENDPOINT>
----

> === Runtime Environment Updates

Reviewer 2: Update Runtime Environment

> * Make sure all environments are running on Mule 4.9.0+

Reviewer 2: Numbered list
Periods

* Verify all environments are using JDK 17+
* Make sure Neo4j servers are version 4.4+ or 5.x

> === Metadata Refresh

Reviewer 2: Refresh Metadata

> * Regenerate metadata in Anypoint Studio for improved design-time experience

Reviewer 2: Numbered list 
periods

* Test the connector operations to ensure metadata is properly resolved

> * Validate DataWeave expressions benefit from enhanced type information

Reviewer 2: Validate that DataWeave expressions?

(It's hard to parse without "that")

=== Verify the Upgrade

After you install the latest version of the connector, verify the upgrade:

> * Make sure no REST URL references remain in configurations

Reviewer 2: Numbered list
periods

* Verify the `pom.xml` file shows version 4.0.0 of the Neo4j Connector
* Use the *Test Connection* feature to validate BOLT connectivity
* Check that the operations show comprehensive field suggestions in Studio
* Execute existing unit tests to ensure functionality is preserved
* Verify no deprecation warnings appear during application startup

== Troubleshooting

> If you encounter issues, perform these troubleshooting guidelines.

Reviewer 2: guidelines? Maybe "steps"?

=== Troubleshoot Connection Failures After Upgrade

If the connector cannot connect to Neo4j server after upgrade, follow these steps:

> . Verify that the Neo4j server is version 4.4+ (Neo4j 3.x is no longer supported)

Reviewer 2: Periods

. Make sure the BOLT protocol is enabled on Neo4j server

> . Remove any REST URL references from configuration

Reviewer 2: ```suggestion
. Remove any REST URL references from the configuration.
```

> Error example: +

Reviewer 2: Is this part of the previous step?

`Neo4j.ClientError.Security.Unauthorized`

> Therefore, verify credentials and server accessibility via BOLT port (default 7687).

Reviewer 2: Should this be another step, rather than "Therefore"?

=== Troubleshoot Metadata Not Displaying in Studio

If the connector operations show empty metadata or missing field suggestions, try these steps to solve the issue:

> * Ensure sample data exists in your Neo4j database for the labels you're using

Reviewer 2: Numbered list
Periods

* Regenerate metadata by refreshing the connector in Studio
* Verify the BOLT connection is working properly

=== Troubleshoot Runtime Compatibility Issues

> If the Mule application fails to start with JDK or Mule Runtime errors, try these steps:

Reviewer 2: ```suggestion
If the Mule application fails to start with JDK or Mule runtime errors, try these steps:
```

> . Upgrade to JDK 17+ (JDK 8/11 no longer supported)

Reviewer 2: Periods

. Upgrade to Mule Runtime 4.9.0+ (earlier versions incompatible)
. Update Maven build configuration for JDK 17 compatibility

> === Reverting the Upgrade

Reviewer 2: ```suggestion
=== Revert the Upgrade
```

If you need to revert to the previous version of Neo4j Connector, go to Anypoint Studio and in the project’s `pom.xml` file, change the `mule-neo4j-connector` dependency version 3.0.0 to the previous version.

> [NOTE]

Reviewer 2: Should this be IMPORTANT?
